### PR TITLE
Platform default parameters read from qibolab platforms

### LINKED ISSUE #594: Nshots defaults to None
I thing this may be related to routines like rabi amplitude, which uses nshots to compute stuff, crashing. This should default to the nshots parameters on the platform.

```py

DEFAULT_PARENT_PARAMETERS = {
    "nshots": None,
    "relaxation_time": None,
}
"""Default values of the parameters of `Parameters`"""


```

### PR DESCRIPTION
**Pull request by GabrielePalazzo**
Closes #594 

Parameters not explicitly defined in the qibocal runcard are read from qibolab runcards.

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.
- [ ] Compatibility with Qibo modules (Please edit this section if the current pull request is not compatible with the following branches).
    - [ ] Qibo: `master`
    - [ ] Qibolab: `main`
    - [ ] Qibolab_platforms_qrc: `main`

### CODE REVIEW

File: src/qibocal/auto/task.py [hunk 1]
```diff
@@ -134,6 +135,18 @@ def run(
...
+        parameters_keys = set(
+            self.operation.parameters_type.__dict__["__annotations__"].keys()
+        ) | set(Parameters.__dict__["__annotations__"].keys())
+        # read all settings defined in platform-specific qibolab runcard
+        for setting, value in platform.settings.__dict__.items():
+            # save values non defined in qibocal runcard,
+            # but that are used by the class
+            if setting not in self.action.parameters:
+                if setting in parameters_keys:
+                    self.action.parameters[setting] = value
```
Comment (andrea-pasquale) [inline src/qibocal/auto/task.py:148]: I think that this can be done more easily...
What about this?
```suggestion
        if self.parameters.nshots is None:
            self.parameters.nshots = platform.settings.nshots

        if self.parameters.relaxation_time is None:
            self.parameters.relaxation_time = platform.settings.relaxation_time
```
I tested and it seems to work as expected.
You can also remove the deepcopy that you did before I believe.

**Review by andrea-pasquale — COMMENTED** [review]
Thanks @GabrielePalazzo for implementing this.
Your solution works but it is a bit convoluted.
I put in a comment another possible solution which seems easier to implement.
Let me know if it works also for you.

File: src/qibocal/auto/task.py (hunk 1 above)
Comment (GabrielePalazzo) [reply src/qibocal/auto/task.py:148]: My only concern with this solution is that if in the future we add more parameters to `qibolab` runcards, it is going to pass under the radar, since this behaviour is not well tested. 
By the way, I’m not satisfied with the tests I drafted yesterday, so I need to think some way to implement some testing.

Comment (andrea-pasquale) [reply src/qibocal/auto/task.py:148]: Currently I believe that the only common parameters between all the routines are `nshots` and `relaxation_time`.

Comment (Jacfomg) [reply src/qibocal/auto/task.py:148]: For testing, we just need to add some routines in the test runcard settings those fields to none, `rabi_amplitude` would be one that could work as it uses `params.nshots` for the fit.

Comment (Edoardo-Pedicillo) [inline src/qibocal/auto/task.py]:
```suggestion
for setting in parameters_keys:
            if setting not in self.action.parameters:
                if hasattr(platform.settings, setting):
                    self.action.parameters[setting]=getattr(platform.settings, setting)

```
I like your solution. I understand that for a few parameters, maybe the @andrea-pasquale approach is better, but this is more general. I suggest not using the dictionaries for better readability.

Comment (Edoardo-Pedicillo) [reply src/qibocal/auto/task.py]: I know that maybe it is too late but if @andrea-pasquale agrees I prefer to restore your solution, in this way it is easier to add default parameters. With my suggested changes (they should work) the only convoluted part is the `parameters_keys` definition, I didn't find a better solution for now.

Comment (GabrielePalazzo) [reply src/qibocal/auto/task.py]: I agree. I'll commit your suggestion as soon as he replies.

Comment (Jacfomg) [reply src/qibocal/auto/task.py]: I would go with @andrea-pasquale solution as it is the one I already implemented in qibolab and I suggested the nshots and relaxation_time as parameters for all routines and I could thing of any other that needs to be shared.

Comment (andrea-pasquale) [reply src/qibocal/auto/task.py]: To be more specific this approach will only work with attributes in `platform.settings` which are only 3 at the moment:
https://github.com/qiboteam/qibolab/blob/53737e12aa7400b22e100eabf538fab9a77667d6/src/qibolab/platform.py#L23-L32
And the sampling rate should not be there I believe.

Comment (Jacfomg) [reply src/qibocal/auto/task.py]: Also, maybe before discussing which one to use we should get one that passes tests, right ?

Comment (GabrielePalazzo) [reply src/qibocal/auto/task.py]: Ok, @andrea-pasquale's solution doesn't work. @Edoardo-Pedicillo's and mine work.
The problem is that `self.parameters()` calls `<SomeChildParameterClass>.load()` which pops elements from its input argument.

I think we should definitely avoid changing the value of function arguments, e.g.:
```
    @classmethod
    def load(cls, input_parameters):
        """Load parameters from runcard.

        Possibly looking into previous steps outputs.
        Parameters defined in Parameters class are removed from `parameters`
        before `cls` is created.
        Then `nshots` and `relaxation_time` are assigned to cls.

        .. todo::

            move the implementation to History, since it is required to resolve
            the linked outputs

        """
        default_parent_parameters = deepcopy(DEFAULT_PARENT_PARAMETERS)
        parameters = deepcopy(input_parameters)
        for parameter, value in default_parent_parameters.items():
            default_parent_parameters[parameter] = parameters.pop(parameter, value)
        instantiated_class = cls(**parameters)
        for parameter, value in default_parent_parameters.items():
            setattr(instantiated_class, parameter, value)
        return instantiated_class
```

Comment (GabrielePalazzo) [reply src/qibocal/auto/task.py]: With the change in `Parameters.load()` all solutions pass test locally on dummy, so it's just a matter of how the code is self-explainable and robust to future changes. I'll leave it up to you.

**Comment by andrea-pasquale** [comment]
Thanks for all the updates @GabrielePalazzo.
As soon as you fixed tests we should be ready to merge I think.

**Comment by GabrielePalazzo** [comment]
> Thanks for all the updates @GabrielePalazzo. As soon as you fixed tests we should be ready to merge I think.

Apparently, `test_auto.py` fails because some runcards (e.g.: `vertical.yaml`) do not have `operation` (required by `self.parameters`). Moving the new lines to the `try`/`except` block makes tests pass, but I don't know if you like this solution.

**Comment by andrea-pasquale** [comment]
> Apparently, `test_auto.py` fails because some runcards (e.g.: `vertical.yaml`) do not have `operation` (required by `self.parameters`). Moving the new lines to the `try`/`except` block makes tests pass, but I don't know if you like this solution.
 We are already doing something like this here 
https://github.com/qiboteam/qibocal/blob/578f07b1c20288564fe91f533afe68225de9a499/src/qibocal/auto/task.py#L137-L142C26
So I guess that it is fine.

File: src/qibocal/auto/task.py [hunk 2]
```diff
@@ -135,6 +135,13 @@ def run(
         task_qubits = self._allocate_local_qubits(qubits, platform)
 
         try:
+            if self.parameters.nshots is None:
+                print(self.action)
```
Comment (Edoardo-Pedicillo) [inline src/qibocal/auto/task.py:139]:
```suggestion
```

**Review by Edoardo-Pedicillo — APPROVED** [review]
LGTM

**Review by Jacfomg — APPROVED** [review]

**Review by andrea-pasquale — APPROVED** [review]
